### ISSUE TITLE
[Feature request] Compatibility with "Railcraft Reborn" steam

### ISSUE THREAD

**Issue by shade1822226911** [AztechMC/Modern-Industrialization#827]
I'd like compatibility with Railcraft via making modern industrialization able to use railcraft steam since they're both fluids by forge standards. You could also make it an config option so people don't skip the mi boilers

**Comment by Technici4n** [comment]
It's a bit complicated. Boilers are locked to only accept steam, and this locking system only accepts a single fluid. It is difficult to accept multiple versions of steam.

**Comment by hron84** [comment]
In this case, can you make the steam inputs to not accept the fluid at all? Because the current situation is the machines are do accepting Railcraft steam as fluid just does not _work_ with it, which causes a confusing situation.

Alternative suggestion is a - cheap - fluid adapter block or cover, that can accept various fluids categorized as steam and converts them directly to MI steam. It could mean BigReactor's steam, Mek's steam, whatever, AFAIK there is a whole tagging system for liquids that can be detected and converted to steam.

**Comment by Swedz** [comment]
Blocking certain fluids from being used in machines would be even more confusing. MI likely will not add new blocks to convert steam fluids. That would be better fitting for some standalone mod.

**Comment by hron84** [comment]
From player perspective, if the machine does not accept the fluid I pipe into it, it means it could not use that fluid. However, especially in early game, every mB of fluids are valuable, piping a bunch of fluids to a machine to just watch it does nothing with that is more confusing I think than not accepting the fluid at all. If you do not limit the fluids on the machine at all, someone would pipe - accidentally - molten diamond into the machine and will not notice that. I do not know how you implemented "it can accept only single fluid" but you should consider some more restrictive filtering not in recipes but in the tank level itself.

Also, notifications, warnings should be written to places because from player perspective, there is no difference between RC steam and MI steam and for example, Mek steam. It is just steam, created from water by heating it.

If you meant to fuel your machines your own fluid, you should consider to rename it something "overheated water" or something that makes clear you did not meant "steam" - since other steams are not considered as fuel.

**Comment by dan96kid** [comment]
Just chiming in to add my support for this idea.

Been playing around in a big modpack (All The Mods 10) that includes this + other mods. I find it annoying that MI machines like the bronze-tier stuff refuse to work with steam from other mods.

**Comment by Kazuhiko-Gushiken** [comment]
> Just chiming in to add my support for this idea.
> 
> Been playing around in a big modpack (All The Mods 10) that includes this + other mods. I find it annoying that MI machines like the bronze-tier stuff refuse to work with steam from other mods.

The same reason that FE cannot be used with MI machines, it isolates power generation in all forms to MI alone. Allowing MI machines/generators to use other mods' steam would be the same bi-directional compat that is disabled by default

**Comment by LordMZTE** [comment]
We implemented this for our modpack using a mixin by converting any steam into MI steam when inserted into an MI machine: https://git.tilera.org/Anvilcraft/anvilcraft-core/src/commit/854ce3f669aca2bf0778aabb6d110c1b1acaf461/src/main/java/net/anvilcraft/core/mixin/MixinMIFluidStorageFluidHandler.java

Why not do something similar upstream, perhaps gated behind a config option?

**Comment by Technici4n** [comment]
@LordMZTE yeah that's a clever approach and we might upstream this. Thank you!

**Comment by Swedz** [comment]
I'm not sure that approach accounts for slots that are auto locked to steam such as boilers and bronze/steel machines

**Comment by LordMZTE** [comment]
> I'm not sure that approach accounts for slots that are auto locked to steam such as boilers and bronze/steel machines

I have tested this with a steel machine and it seems to work fine with Railcraft steam. (Not with slots that aren't always steam though, now that you mention it :P)

**Comment by Swedz** [comment]
I read the mixin code very briefly on my phone so I didn't quite comprehend that it was converting any steam to MI steam. That makes more sense.

It may not matter at any point (since its unlikely the code would change), but it's worth mentioning that the `@Overwrite` mixin annotation is not very robust in terms of maintainability on your end. If anything changes in that method upstream, you would need to reflect those changes or fall behind, possibly resulting in a bug. I'd recommend using `@ModifyVariable` on the `FluidStack resource` parameter instead.

I'd like to see configurable stacks to be able to have a filter (defined by the machine, not the user) that allows it to accept a variety of inputs. Rather than just locking it to a single item or fluid. That way we could make boilers and steam machines have a locked slot that outright accepts any steam without needing to do conversions.

**Comment by Swedz** [comment]
Balance is also a concern, since to my understanding Railcraft steam is significantly cheaper and easier to make than MI steam. I suppose it wouldn't be all that different than the bidirectional energy compat config option.
